File: 2024/day06/php/solution.php
<?php

function parseGrid() {
    $input = file_get_contents(__DIR__ . '/../input.txt');
    $lines = explode("\n", trim($input));

    // Parse the grid
    $grid = [];
    $startRow = 0;
    $startCol = 0;
    $startDir = '';

    foreach ($lines as $i => $line) {
        $grid[] = str_split($line);
        for ($j = 0; $j < strlen($line); $j++) {
            $char = $line[$j];
            if (in_array($char, ['^', 'v', '<', '>'])) {
                $startRow = $i;
                $startCol = $j;
                $startDir = $char;
            }
        }
    }

    return [$grid, $startRow, $startCol, $startDir];
}

function simulateGuard($grid, $startRow, $startCol, $startDir, $detectLoop = false) {
    $height = count($grid);
    $width = count($grid[0]);

    // Direction vectors: up, right, down, left
    $directions = [
        '^' => [-1, 0],
        '>' => [0, 1],
        'v' => [1, 0],
        '<' => [0, -1]
    ];

    // Turn right mapping
    $turnRight = [
        '^' => '>',
        '>' => 'v',
        'v' => '<',
        '<' => '^'
    ];

    // Track visited positions
    $visited = [];
    $visitedKey = $startRow . ',' . $startCol;
    $visited[$visitedKey] = true;

    // Track states (position + direction) for loop detection
    $states = [];
    if ($detectLoop) {
        $stateKey = $startRow . ',' . $startCol . ',' . $startDir;
        $states[$stateKey] = true;
    }

    $row = $startRow;
    $col = $startCol;
    $dir = $startDir;

    // Simulate guard movement
    while (true) {
        $delta = $directions[$dir];
        $nextRow = $row + $delta[0];
        $nextCol = $col + $delta[1];

        // Check if guard leaves the map
        if ($nextRow < 0 || $nextRow >= $height || $nextCol < 0 || $nextCol >= $width) {
            return $detectLoop ? false : $visited;
        }

        // Check if there's an obstacle ahead
        if ($grid[$nextRow][$nextCol] === '#') {
            // Turn right
            $dir = $turnRight[$dir];

            // Check for loop after turning
            if ($detectLoop) {
                $stateKey = $row . ',' . $col . ',' . $dir;
                if (isset($states[$stateKey])) {
                    return true; // Loop detected
                }
                $states[$stateKey] = true;
            }
        } else {
            // Move forward
            $row = $nextRow;
            $col = $nextCol;
            $visitedKey = $row . ',' . $col;
            $visited[$visitedKey] = true;

            // Check for loop (same position and direction)
            if ($detectLoop) {
                $stateKey = $row . ',' . $col . ',' . $dir;
                if (isset($states[$stateKey])) {
                    return true; // Loop detected
                }
                $states[$stateKey] = true;
            }
        }
    }
}

function part1() {
    list($grid, $startRow, $startCol, $startDir) = parseGrid();
    $visited = simulateGuard($grid, $startRow, $startCol, $startDir, false);
    return count($visited);
}

function part2() {
    list($grid, $startRow, $startCol, $startDir) = parseGrid();

    $height = count($grid);
    $width = count($grid[0]);

    // First, get all positions visited in the normal path
    $normalVisited = simulateGuard($grid, $startRow, $startCol, $startDir, false);

    $loopCount = 0;

    // Try placing an obstruction at each position (except the starting position)
    foreach ($normalVisited as $posKey => $value) {
        list($testRow, $testCol) = explode(',', $posKey);
        $testRow = intval($testRow);
        $testCol = intval($testCol);

        // Skip the starting position
        if ($testRow === $startRow && $testCol === $startCol) {
            continue;
        }

        // Skip if already an obstacle
        if ($grid[$testRow][$testCol] === '#') {
            continue;
        }

        // Temporarily place an obstruction
        $grid[$testRow][$testCol] = '#';

        // Simulate and check for loop
        $hasLoop = simulateGuard($grid, $startRow, $startCol, $startDir, true);

        if ($hasLoop) {
            $loopCount++;
        }

        // Remove the obstruction
        $grid[$testRow][$testCol] = '.';
    }

    return $loopCount;
}

echo "Part 1: " . part1() . "\n";
echo "Part 2: " . part2() . "\n";

?>

File: 2024/day06/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Solution {
    // Direction indices: 0=Up, 1=Right, 2=Down, 3=Left
    private static final int[] DR = {-1, 0, 1, 0};
    private static final int[] DC = {0, 1, 0, -1};

    private static int rows, cols;
    private static char[][] grid;
    private static int startRow, startCol, startDir;

    // Flat array for state tracking - better cache performance
    // Index = (row * cols + col) * 4 + dir
    private static int[] stateVersion;
    private static int currentVersion = 0;

    // Separate array for tracking original visited positions
    private static boolean[][] originalPath;
    private static int part1Answer = -1;

    private static void parseGrid(List<String> lines) {
        rows = lines.size();
        cols = lines.get(0).length();
        grid = new char[rows][cols];
        stateVersion = new int[rows * cols * 4];
        originalPath = new boolean[rows][cols];
        currentVersion = 0;
        part1Answer = -1;

        for (int r = 0; r < rows; r++) {
            String line = lines.get(r);
            for (int c = 0; c < line.length(); c++) {
                char ch = line.charAt(c);
                grid[r][c] = ch;

                if (ch == '^') {
                    startRow = r; startCol = c; startDir = 0;
                    grid[r][c] = '.';
                } else if (ch == '>') {
                    startRow = r; startCol = c; startDir = 1;
                    grid[r][c] = '.';
                } else if (ch == 'v') {
                    startRow = r; startCol = c; startDir = 2;
                    grid[r][c] = '.';
                } else if (ch == '<') {
                    startRow = r; startCol = c; startDir = 3;
                    grid[r][c] = '.';
                }
            }
        }
    }

    // Get original visited path (for part1 and part2 optimization)
    private static int getOriginalPath() {
        if (part1Answer >= 0) return part1Answer;

        int row = startRow, col = startCol, dir = startDir;
        int count = 0;

        while (true) {
            if (!originalPath[row][col]) {
                originalPath[row][col] = true;
                count++;
            }

            int nextRow = row + DR[dir];
            int nextCol = col + DC[dir];

            if (nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols) {
                part1Answer = count;
                return count;
            }

            if (grid[nextRow][nextCol] == '#') {
                dir = (dir + 1) & 3;
            } else {
                row = nextRow;
                col = nextCol;
            }
        }
    }

    // Simulate with obstruction, returns true if loop detected
    private static boolean detectsLoop(int obsRow, int obsCol) {
        // Increment version to "clear" stateVersion array in O(1)
        currentVersion++;

        int row = startRow, col = startCol, dir = startDir;

        while (true) {
            // Flat index calculation
            int stateIdx = (row * cols + col) * 4 + dir;

            // Loop detection check using version comparison
            if (stateVersion[stateIdx] == currentVersion) {
                return true;  // Loop detected
            }
            stateVersion[stateIdx] = currentVersion;

            int nextRow = row + DR[dir];
            int nextCol = col + DC[dir];

            // Check bounds
            if (nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols) {
                return false;  // Guard exits map
            }

            // Check obstacle (including temporary obstruction)
            if (grid[nextRow][nextCol] == '#' || (nextRow == obsRow && nextCol == obsCol)) {
                dir = (dir + 1) & 3;  // Turn right
            } else {
                row = nextRow;
                col = nextCol;
            }
        }
    }

    public static int part1(List<String> lines) {
        parseGrid(lines);
        return getOriginalPath();
    }

    public static int part2(List<String> lines) {
        // Only reparse if needed (allows sharing with part1)
        if (grid == null || grid.length != lines.size()) {
            parseGrid(lines);
        }

        // First get visited positions from unobstructed path
        getOriginalPath();

        int loopCount = 0;

        // Try placing obstruction at each visited position (except start)
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                // Check if position was visited in original path
                if (!originalPath[r][c]) continue;
                if (r == startRow && c == startCol) continue;

                if (detectsLoop(r, c)) {
                    loopCount++;
                }
            }
        }

        return loopCount;
    }

    public static void main(String[] args) throws IOException {
        List<String> lines = Files.readAllLines(Paths.get("../input.txt"));

        // Parse once, use for both parts
        parseGrid(lines);

        int result1 = getOriginalPath();
        System.out.println("Part 1: " + result1);

        int result2 = part2(lines);
        System.out.println("Part 2: " + result2);
    }
}
